File: web_front/src/lib/api/interview.ts
/** 面试模块API客户端，对接后端 /api/v1/interviews 端点（面试流程功能文档 §3.4）。 */

import request from '@/lib/request';
import { TAB_ID } from '@/lib/tabId';

/** 面试类型：1-完整面试 2-快速面试 */
export type InterviewType = 1 | 2;

/** 面试状态：0-进行中 1-已完成 2-已中断 */
export type InterviewStatus = 0 | 1 | 2;

/** 面试阶段（后端 Checkpoint phase） */
export type InterviewPhase = 'not_started' | 'answering' | 'analyzing' | 'summarizing' | 'completed' | 'aborted';

/** 当前题目（后端 QuestionOut） */
export interface ApiInterviewQuestion {
  question_index: number;
  question_no: number;
  question_id: number;
  question_text: string;
  question_type: number;
  category: number | null;
  is_follow_up: boolean;
}

/** 创建面试响应（后端 InterviewCreateResponse） */
export interface ApiInterviewCreateResponse {
  interview_id: number;
  epoch: number;
  status: number;
  type: number;
  total_questions: number;
  current_question: ApiInterviewQuestion | null;
}

/** 面试当前状态响应（后端 InterviewStateResponse） */
export interface ApiInterviewState {
  interview_id: number;
  status: InterviewStatus;
  type: number;
  phase: InterviewPhase;
  question_index: number;
  epoch: number;
  answered_count: number;
  total_questions: number;
  current_question: ApiInterviewQuestion | null;
}

/** 单题分析摘要（后端 AnswerAnalysisOut） */
export interface ApiAnswerAnalysis {
  score: number;
  comment: string;
  correctness: string;
  technical_depth: number;
  completeness: number;
  logic: number;
  key_points: string[];
  weaknesses: string[];
}

/** 提交回答响应（后端 AnswerSubmitResponse） */
export interface ApiAnswerSubmitResponse {
  interview_id: number;
  question_index: number;
  analysis: ApiAnswerAnalysis;
  duplicated: boolean;
  phase: 'answering' | 'summarizing';
  next_question: ApiInterviewQuestion | null;
}

/** 面试报告（后端 InterviewReportResponse） */
export interface ApiInterviewReport {
  interview_id: number;
  total_score: number;
  dimension_scores: Record<string, number> | null;
  strengths: string[];
  weaknesses: string[];
  capability_profile: Record<string, string> | null;
  suggestions: string[];
  summary: string;
  question_count: number;
  follow_up_count: number;
  total_duration: number | null;
  created_at: string;
}

/** 报告状态响应（后端 ReportStatusResponse） */
export interface ApiReportStatus {
  status: 'generating' | 'ready' | 'failed' | 'invalid';
  report: ApiInterviewReport | null;
}

/** 面试记录列表项（后端 InterviewListItem） */
export interface ApiInterviewListItem {
  interview_id: number;
  status: InterviewStatus;
  type: number;
  total_score: number | null;
  follow_up_count: number;
  question_count: number;
  answered_count: number;
  report_ready: boolean;
  /** 是否已正式启动（False=草稿，设备检测前；历史页据此分流跳转） */
  is_started: boolean;
  created_at: string;
  interview_time: string | null;
  total_duration: number | null;
}

/** 面试记录列表响应 */
export interface ApiInterviewListResponse {
  items: ApiInterviewListItem[];
  total: number;
  page: number;
  page_size: number;
}

/** 面试统计响应（GET /interviews/stats，含软删除记录，删除不影响平均分） */
export interface ApiInterviewStats {
  /** 可见（未删除）记录总数，用于历史页分页 */
  total: number;
  /** 已完成面试次数（含已删记录） */
  completed_count: number;
  /** 已完成面试平均分（含已删记录，无则 null） */
  avg_score: number | null;
}

/** 逐题详情（后端 InterviewQuestionDetail，仅已结束面试返回） */
export interface ApiInterviewQuestionDetail {
  question_index: number;
  question_no: number;
  question_id: number;
  question_text: string;
  question_type: number;
  category: number | null;
  is_follow_up: boolean;
  user_answer: string | null;
  ai_score: number | null;
  ai_comment: string | null;
  answer_duration: number | null;
}

/** 逐题详情列表响应 */
export interface ApiInterviewQuestionListResponse {
  items: ApiInterviewQuestionDetail[];
  total: number;
}

/** 后端409冲突响应体：{reason, latest_state} 或 {code, message} */
export interface ApiConflictDetail {
  reason?: string;
  latest_state?: ApiInterviewState | null;
  code?: string;
  message?: string;
}

/** 题型常量 → 文案映射（对齐后端 question_type） */
export const QUESTION_TYPE_LABEL: Record<number, string> = {
  1: '技术题',
  2: '项目题',
  3: '行为题',
};

/** 题目维度常量 → 文案映射（对齐后端 category） */
export const CATEGORY_LABEL: Record<number, string> = {
  1: '技术八股',
  2: '项目与社会实践',
  3: '综合素养',
  4: '架构设计',
};

/** 面试类型 → 文案映射 */
export const INTERVIEW_TYPE_LABEL: Record<number, string> = {
  1: '完整面试',
  2: '快速面试',
};

/**
 * 从 axios 错误中提取后端409冲突详情。
 * @param err 任意抛出的错误对象
 */
export function extractConflict(err: unknown): ApiConflictDetail | null {
  const detail = (err as { response?: { status?: number; data?: { detail?: unknown } } })?.response?.data?.detail;
  if (detail && typeof detail === 'object') return detail as ApiConflictDetail;
  return null;
}

/**
 * 创建面试会话：简历状态硬校验 + 预生成基础题（真实LLM，耗时5~20秒）。
 * @param resumeId 简历ID（须已就绪）
 * @param type 面试类型 1-完整 2-快速
 */
export async function createInterview(resumeId: number, type: InterviewType): Promise<ApiInterviewCreateResponse> {
  const { data } = await request.post<ApiInterviewCreateResponse>(
    '/interviews',
    { resume_id: resumeId, type, tab_id: TAB_ID },
    { timeout: 120000 },
  );
  return data;
}

/**
 * 查询面试当前状态（刷新恢复/超时兜底轮询）。
 * 携带 tab_id 时触发客户端租约激活（同页幂等/新页接管 epoch+1）。
 * @param interviewId 面试会话ID
 * @param withTab 是否携带本标签页标识（进入/刷新面试页时 true）
 */
export async function getInterviewState(
  interviewId: number,
  withTab = false,
): Promise<ApiInterviewState> {
  const { data } = await request.get<ApiInterviewState>(`/interviews/${interviewId}`, {
    params: withTab ? { tab_id: TAB_ID } : undefined,
    timeout: 15000,
  });
  return data;
}

/**
 * 设备检测通过后正式启动面试（后端草稿态 not_started → answering，幂等）。
 * @param interviewId 面试会话ID
 */
export async function startInterview(interviewId: number): Promise<ApiInterviewState> {
  const { data } = await request.post<ApiInterviewState>(
    `/interviews/${interviewId}/start`,
    null,
    { timeout: 15000 },
  );
  return data;
}

/**
 * 提交回答：Fast Decision 即时判定，秒级返回下一题（v2，全量分析走异步 Worker）。
 * 幂等键 (interviewId, questionIndex)，重试安全。
 * @param interviewId 面试会话ID
 * @param questionIndex 所答题目题序（状态版本token）
 * @param answer 回答文本（语音转写或键盘输入）
 * @param tabEpoch 客户端租约epoch
 * @param answerDuration 回答时长（秒，可选）
 */
export async function submitAnswer(
  interviewId: number,
  questionIndex: number,
  answer: string,
  tabEpoch: number,
  answerDuration?: number,
): Promise<ApiAnswerSubmitResponse> {
  const { data } = await request.post<ApiAnswerSubmitResponse>(
    `/interviews/${interviewId}/answers`,
    {
      question_index: questionIndex,
      answer,
      tab_epoch: tabEpoch,
      ...(answerDuration !== undefined ? { answer_duration: answerDuration } : {}),
    },
    { timeout: 30000 },
  );
  return data;
}

/**
 * 获取面试报告（未生成时返回 generating 并惰性触发一次生成）。
 * @param interviewId 面试会话ID
 */
export async function getInterviewReport(interviewId: number): Promise<ApiReportStatus> {
  const { data } = await request.get<ApiReportStatus>(`/interviews/${interviewId}/report`, {
    timeout: 15000,
  });
  return data;
}

/**
 * 报告生成失败后的手动重试。
 * @param interviewId 面试会话ID
 */
export async function regenerateReport(interviewId: number): Promise<ApiReportStatus> {
  const { data } = await request.post<ApiReportStatus>(
    `/interviews/${interviewId}/report/regenerate`,
    null,
    { timeout: 15000 },
  );
  return data;
}

/**
 * 主动放弃面试（status=2，已答数据保留）。
 * @param interviewId 面试会话ID
 * @param tabEpoch 客户端租约epoch
 */
export async function abortInterview(interviewId: number, tabEpoch: number): Promise<void> {
  await request.post(`/interviews/${interviewId}/abort`, { tab_epoch: tabEpoch }, { timeout: 15000 });
}

/**
 * 删除面试记录（软删除：草稿/进行中/已中断/已完成均可删，不影响统计）。
 * @param interviewId 面试会话ID
 */
export async function deleteInterview(interviewId: number): Promise<void> {
  await request.delete(`/interviews/${interviewId}`, { timeout: 15000 });
}

/**
 * 查询面试统计（总次数/完成数/平均分，含软删除记录——删除不影响平均分）。
 */
export async function getInterviewStats(): Promise<ApiInterviewStats> {
  const { data } = await request.get<ApiInterviewStats>('/interviews/stats', { timeout: 15000 });
  return data;
}

/**
 * 分页查询我的面试记录列表（历史页）。
 * @param page 页码
 * @param pageSize 页大小
 */
export async function getInterviewList(page = 1, pageSize = 20): Promise<ApiInterviewListResponse> {
  const { data } = await request.get<ApiInterviewListResponse>('/interviews', {
    params: { page, page_size: pageSize },
    timeout: 15000,
  });
  return data;
}

/**
 * 查询已结束面试的逐题详情（报告页逐题展示；进行中面试返回409）。
 * @param interviewId 面试会话ID
 */
export async function getInterviewQuestions(
  interviewId: number,
): Promise<ApiInterviewQuestionListResponse> {
  const { data } = await request.get<ApiInterviewQuestionListResponse>(
    `/interviews/${interviewId}/questions`,
    { timeout: 15000 },
  );
  return data;
}

File: api_server/app/schemas/interview.py
"""面试相关请求/响应模型（文档 §3.4 API 契约）。"""

from datetime import datetime

from pydantic import BaseModel, Field


class InterviewCreateRequest(BaseModel):
    """创建面试请求模型（§3.1）。"""

    resume_id: int = Field(..., ge=1, description="使用的简历ID")
    type: int = Field(1, ge=1, le=2, description="面试类型 1-完整面试 2-快速面试")
    tab_id: str = Field(..., min_length=1, max_length=64, description="创建面试的标签页唯一标识")


class QuestionOut(BaseModel):
    """当前题目输出模型（未完成面试仅返回当前题，§7.2）。"""

    question_index: int = Field(..., description="发问顺序题序（1起，含追问）")
    question_no: int = Field(..., description="基础题号（追问题与父题同号）")
    question_id: int = Field(..., description="题目ID")
    question_text: str = Field(..., description="题目内容")
    question_type: int = Field(..., description="题型 1-技术题 2-项目题 3-行为题")
    category: int | None = Field(None, description="维度 1-技术基础 2-项目经验 3-综合素质 4-架构设计")
    is_follow_up: bool = Field(False, description="是否追问题")


class InterviewCreateResponse(BaseModel):
    """创建面试响应模型（返回面试 id + epoch + 首题，§3.4）。"""

    interview_id: int = Field(..., description="面试会话ID（即 session_id）")
    epoch: int = Field(..., description="当前客户端持有的租约 epoch")
    status: int = Field(..., description="面试状态 0-进行中")
    type: int = Field(..., description="面试类型")
    total_questions: int = Field(..., description="基础题总数")
    current_question: QuestionOut | None = Field(None, description="首题")


class InterviewStateResponse(BaseModel):
    """面试当前状态响应模型（刷新恢复/超时兜底轮询，§3.4）。"""

    interview_id: int = Field(..., description="面试会话ID")
    status: int = Field(..., description="面试状态 0-进行中 1-已完成 2-已中断")
    type: int = Field(..., description="面试类型")
    phase: str = Field(..., description="阶段 not_started/answering/analyzing/summarizing/completed/aborted")
    question_index: int = Field(..., description="当前题序（状态版本token，§5.5）")
    epoch: int = Field(..., description="当前客户端租约 epoch（GET 携带 tab_id 时可能+1）")
    answered_count: int = Field(0, description="已答题数（含追问）")
    total_questions: int = Field(0, description="基础题总数")
    current_question: QuestionOut | None = Field(None, description="当前题目")


class AnswerSubmitRequest(BaseModel):
    """提交回答请求模型（§8.4）。"""

    question_index: int = Field(..., ge=1, description="所答题目题序（状态版本token）")
    answer: str = Field(..., min_length=1, max_length=10000, description="回答文本（语音转写或键盘输入）")
    tab_epoch: int = Field(..., ge=1, description="客户端租约 epoch（双开裁决）")
    answer_duration: int | None = Field(None, ge=0, le=3600, description="回答时长（秒）")


class AnswerAnalysisOut(BaseModel):
    """单题分析摘要输出模型（§9.2 合并输出的展示子集）。"""

    score: int = Field(..., description="综合评分 1-5")
    comment: str = Field(..., description="综合评价")
    correctness: str = Field("", description="正确性简述")
    technical_depth: int = Field(1, description="技术深度 1-5")
    completeness: int = Field(1, description="完整性 1-5")
    logic: int = Field(1, description="逻辑性 1-5")
    key_points: list[str] = Field(default_factory=list, description="回答要点")
    weaknesses: list[str] = Field(default_factory=list, description="薄弱点")


class AnswerSubmitResponse(BaseModel):
    """提交回答响应模型（分析摘要+评分+下一题或summarizing，§3.4）。"""

    interview_id: int = Field(..., description="面试会话ID")
    question_index: int = Field(..., description="本次已答题目题序")
    analysis: AnswerAnalysisOut = Field(..., description="本题主观分析摘要与评分")
    duplicated: bool = Field(False, description="是否命中幂等直接返回既有结果")
    phase: str = Field(..., description="下一阶段 answering/summarizing")
    next_question: QuestionOut | None = Field(None, description="下一题（追问或下一基础题）")


class AbortRequest(BaseModel):
    """主动放弃请求模型。"""

    tab_epoch: int = Field(..., ge=1, description="客户端租约 epoch")


class InterviewReportResponse(BaseModel):
    """面试报告响应模型（§13/§14.3）。"""

    interview_id: int = Field(..., description="面试会话ID")
    total_score: float = Field(..., description="总评分（百分制）")
    dimension_scores: dict[str, float] | None = Field(None, description="各维度得分")
    strengths: list[str] = Field(default_factory=list, description="回答优点")
    weaknesses: list[str] = Field(default_factory=list, description="知识薄弱点")
    capability_profile: dict[str, str] | None = Field(None, description="能力画像")
    suggestions: list[str] = Field(default_factory=list, description="改进建议")
    summary: str = Field(..., description="总评")
    question_count: int = Field(0, description="题目数量")
    follow_up_count: int = Field(0, description="追问次数")
    total_duration: int | None = Field(None, description="面试总时长（秒）")
    created_at: datetime = Field(..., description="报告生成时间")

    model_config = {"from_attributes": True}


class ReportStatusResponse(BaseModel):
    """报告状态响应模型（未生成时返回generating，§13.1）。"""

    status: str = Field(..., description="generating/ready/failed/invalid")
    report: InterviewReportResponse | None = Field(None, description="报告内容（status=ready时返回）")


class InterviewListItem(BaseModel):
    """面试记录列表项模型（历史页展示）。"""

    interview_id: int = Field(..., description="面试会话ID")
    status: int = Field(..., description="面试状态 0-进行中 1-已完成 2-已中断")
    type: int = Field(..., description="面试类型 1-完整 2-快速")
    total_score: float | None = Field(None, description="面试总分（已完成）")
    follow_up_count: int = Field(0, description="追问次数")
    question_count: int = Field(0, description="基础题总数")
    answered_count: int = Field(0, description="已答题数（含追问）")
    report_ready: bool = Field(False, description="报告是否已生成")
    is_started: bool = Field(True, description="是否已正式启动（False=草稿，待设备检测）")
    created_at: datetime = Field(..., description="创建时间")
    interview_time: datetime | None = Field(None, description="完成时间")
    total_duration: int | None = Field(None, description="面试总时长（秒）")


class InterviewListResponse(BaseModel):
    """面试记录列表响应模型（分页）。"""

    items: list[InterviewListItem] = Field(..., description="面试记录列表")
    total: int = Field(..., description="记录总数")
    page: int = Field(..., description="当前页码")
    page_size: int = Field(..., description="页大小")


class InterviewQuestionDetail(BaseModel):
    """逐题详情模型（仅已结束面试返回全量，§7.2约束）。"""

    question_index: int = Field(..., description="发问顺序题序（1起，含追问）")
    question_no: int = Field(..., description="基础题号（追问题与父题同号）")
    question_id: int = Field(..., description="题目ID")
    question_text: str = Field(..., description="题目内容")
    question_type: int = Field(..., description="题型 1-技术题 2-项目题 3-行为题")
    category: int | None = Field(None, description="维度 1-技术基础 2-项目经验 3-综合素质 4-架构设计")
    is_follow_up: bool = Field(False, description="是否追问题")
    user_answer: str | None = Field(None, description="用户回答")
    ai_score: int | None = Field(None, description="AI评分 1-5")
    ai_comment: str | None = Field(None, description="AI评价")
    answer_duration: int | None = Field(None, description="回答时长（秒）")


class InterviewQuestionListResponse(BaseModel):
    """逐题详情列表响应模型。"""

    items: list[InterviewQuestionDetail] = Field(..., description="题目列表（发问顺序）")
    total: int = Field(..., description="题目总数")
